File: leetcode/N-Queens.java
import java.util.*;

public class Solution {
    boolean okay(int r, int c, int n, char[][] m) {
        for (int k = 1; r - k >= 0; ++k) {
            if (m[r - k][c] == 'Q')
                return false;
            if (c - k >= 0 && m[r - k][c - k] == 'Q')
                return false;
            if (c + k < n && m[r - k][c + k] == 'Q')
                return false;
        }
        return true;
    }
    void solve(int r, int n, char[][] m, ArrayList<String[]> res) {
        if (r == n) {
            String[] x = new String[n];
            for (int i = 0; i < n; ++i) x[i] = new String(m[i]);
            res.add(x);
            return;
        }
        for (int i = 0; i < n; ++i) {
            if (okay(r, i, n, m)) {
                m[r][i] = 'Q';
                solve(r + 1, n, m, res);
                m[r][i] = '.';
            }
        }
    }
    public ArrayList<String[]> solveNQueens(int n) {
        ArrayList<String[]> ret = new ArrayList<String[]>();
        char[][] board = new char[n][n];
        for (int i = 0; i < n; ++i)
            for (int j = 0; j < n; ++j)
                board[i][j] = '.';
        solve(0, n, board, ret);
        return ret;
    }

/*    public static void main(String[] args) {
        Solution o = new Solution();
        ArrayList<String[]> a = o.solveNQueens(4);
        ListIterator<String[]> it = a.listIterator();
        while (it.hasNext()) {
            String[] x = it.next();
            for (int i = 0; i < x.length; ++i)
                System.out.println(x[i]);
            System.out.println();
        }
    }*/
}

File: leetcode/N-Queens.cpp
class Solution {
    vector<vector<string> > solutions;
    void place(int n, int k, vector<string> &board) {
        if (k == n) {
            solutions.push_back(board);
            return;
        }
        for (int i = 0; i < n; ++i) {
            if (k > 0) {
                bool ok = true;
                for (int j = 0; j < k; ++j) {
                    if (board[j][i] == 'Q') {  // north
                        ok = false;
                        break;
                    }
                    if (i - (k - j) >= 0 && 
                        board[j][i - (k - j)] == 'Q') {  // northwest
                        ok = false;
                        break;
                    }
                    if (i + (k - j) < n && 
                        board[j][i + (k - j)] == 'Q') {  // northeast
                        ok = false;
                        break;
                    }
                }
                if (!ok)
                    continue;
            }
            board[k][i] = 'Q';
            place(n, k + 1, board);
            board[k][i] = '.';
        }
    }
public:
    vector<vector<string> > solveNQueens(int n) {
        solutions.clear();
        if (n == 0)
            return solutions;
        vector<string> board(n);
        for (int i = 0; i < n; ++i) {
            board[i] = "";
            for (int j = 0; j < n; ++j)
                board[i] += '.';
        }
        place(n, 0, board);
        return solutions;
    }
};
